js/src/api/local/ethkey/worker.js

```diff
@@ -14,68 +14,117 @@
-if (typeof self !== 'object') {
+if (!isWorker) {
   const scope = typeof global === 'undefined' ? window : global;
 
   scope.self = scope;
 }
 
-function bytesToHex (bytes) {
-  return '0x' + Array.from(bytes).map(n => ('0' + n.toString(16)).slice(-2)).join('');
-}
+// keythereum should never be used outside of the browser
+let keythereum = null;
 
-// Logic ported from /ethkey/src/brain.rs
-function phraseToWallet (phrase) {
-  let secret = keccak256.array(phrase);
+if (isWorker) {
+  require('keythereum/dist/keythereum');
 
-  for (let i = 0; i < 16384; i++) {
-    secret = keccak256.array(secret);
+  keythereum = self.keythereum;
+}
+
+function route ({ action, payload }) {
+  if (action in actions) {
+    return actions[action](payload);
   }
 
-  while (true) {
-    secret = keccak256.array(secret);
+  return null;
+}
+
+const actions = {
+  phraseToWallet (phrase) {
+    let secret = keccak256.array(phrase);
+
+    for (let i = 0; i < 16384; i++) {
+      secret = keccak256.array(secret);
+    }
+
+    while (true) {
+      secret = keccak256.array(secret);
 
-    const secretBuf = Buffer.from(secret);
+      const secretBuf = Buffer.from(secret);
 
-    if (secp256k1.privateKeyVerify(secretBuf)) {
-      // No compression, slice out last 64 bytes
-      const publicBuf = secp256k1.publicKeyCreate(secretBuf, false).slice(-64);
-      const address = keccak256.array(publicBuf).slice(12);
+      if (secp256k1.privateKeyVerify(secretBuf)) {
+        // No compression, slice out last 64 bytes
+        const publicBuf = secp256k1.publicKeyCreate(secretBuf, false).slice(-64);
+        const address = keccak256.array(publicBuf).slice(12);
 
-      if (address[0] !== 0) {
-        continue;
+        if (address[0] !== 0) {
+          continue;
+        }
+
+        const wallet = {
+          secret: bytesToHex(secretBuf),
+          public: bytesToHex(publicBuf),
+          address: bytesToHex(address)
+        };
+
+        return wallet;
       }
+    }
+  },
+
+  verifySecret (secret) {
+    const key = Buffer.from(secret.slice(2), 'hex');
+
+    return secp256k1.privateKeyVerify(key);
+  },
+
+  createKeyObject ({ key, password }) {
+    key = Buffer.from(key);
+    password = Buffer.from(password);
+
+    const iv = keythereum.crypto.randomBytes(16);
+    const salt = keythereum.crypto.randomBytes(32);
+    const keyObject = keythereum.dump(password, key, salt, iv);
 
-      const wallet = {
-        secret: bytesToHex(secretBuf),
-        public: bytesToHex(publicBuf),
-        address: bytesToHex(address)
-      };
+    return JSON.stringify(keyObject);
+  },
 
-      return wallet;
+  decryptPrivateKey ({ keyObject, password }) {
+    password = Buffer.from(password);
+
+    try {
+      const key = keythereum.recover(password, keyObject);
+
+      // Convert to array to safely send from the worker
+      return Array.from(key);
+    } catch (e) {
+      return null;
     }
   }
+};
+
+function bytesToHex (bytes) {
```

Comment:
Not 100% sure why this is "duplicated", i.e. not used from `/api/util/format`? (I looked at this the previous round as well, could see a slight difference, but non-duplication is obviously preferred)